module/zfs/bpobj.c

```diff
@@ -242,119 +279,167 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
+
+		err = bpobj_iterate_blkptrs(bpi, func, arg, tx, free);
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			/* push back onto stack for cleanup function */
+			list_insert_head(&stack, bpi);
+			goto out;
 		}
-
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+		/* A subobj with no subobjs below it can be cleaned up. */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
 			}
+			if (free && (bpi->bpi_parent != NULL)) {
+				bpobj_t *p = bpi->bpi_parent->bpi_bpo;
+
+				ASSERT0(bpo->bpo_phys->bpo_num_blkptrs);
+				ASSERT3U(p->bpo_phys->bpo_num_subobjs, >, 0);
+				ASSERT3U(bpi->bpi_index, ==,
+				    p->bpo_phys->bpo_num_subobjs - 1);
+
+				p->bpo_phys->bpo_num_subobjs--;
+
+				VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
+				    p->bpo_phys->bpo_subobjs,
+				    bpi->bpi_index * sizeof (uint64_t),
+				    sizeof (uint64_t), tx));
+
+				/* eliminate the empty subobj */
+				if (bpo->bpo_havesubobj &&
+				    bpo->bpo_phys->bpo_subobjs != 0) {
+					ASSERT0(bpo->bpo_phys->bpo_num_subobjs);
+					err = dmu_object_free(bpo->bpo_os,
+					    bpo->bpo_phys->bpo_subobjs, tx);
+					if (err) {
+						mutex_exit(&bpo->bpo_lock);
+						break;
+					}
+					dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+					bpo->bpo_phys->bpo_subobjs = 0;
+				}
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
+				if (err) {
+					mutex_exit(&bpo->bpo_lock);
+					break;
+				}
+			}
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			continue;
 		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
-		}
+		ASSERT(bpo->bpo_havecomp);
+		/*
+		 * Can't clean up bpobj while it still has subobjs.  Place it
+		 * back in the list ahead of its children.  It will be cleaned
+		 * up once the children are all freed.
+		 */
+		list_insert_head(&stack, bpi);
 
-		bpobj_close(&sublist);
+		/* Add the last subobj to stack. */
```

Comment:
This comment explains what the code below does, which is useful since it's a bit complicated.  I think it would also be useful to have a higher-level comment that explains the overall algorithm, which would explain _why_ we want to add the last subobj to the stack.  I think that's what @behlendorf was getting at with his previous comment: https://github.com/zfsonlinux/zfs/pull/7675/files#r212765323

Reply:
How's this?
```
/*                                                                              
 * Given an initial bpo, start by freeing the BPs that are directly referenced  
 * by that bpo. If the bpo has subobjs, read in its last subobj and add it      
 * to our LIFO stack. By working through our stack, eventually we will encounter
 * a bpo that has no subobjs.  We can free the bpobj_info_t for it, decrement   
 * its parent's subobj count, and if indicated, free the empty bpo from disk.   
 * Parent subobj(s) in our LIFO stack and will need to have their directly      
 * referenced BPs freed and any additonal subobjs processed until they          
 * too have no more subobjs, as so on.                                          
 */     
static int                                                                      
bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,          
    dmu_tx_t *tx, boolean_t free)                                               
{ ```

Reply:
> If the bpo has subobjs, read in its last subobj and add it      
> to our LIFO stack.

You could use the more specific verbs `push` and `pop` when talking about the stack.

Maybe also talk about the fact that we re-push the bpo before pushing its last subobj, and this means that we'll revisit this bpo again after visiting its subobj, and then we will evaluate if we need to repeat this (push last subobj) or if the bpo is now empty.

> We can free the bpobj_info_t for it, decrement   
> its parent's subobj count, and if indicated, free the empty bpo from disk.

I think "indicated" means that `free` is set.  I think the decrement also only happens if `free` is set, so maybe reword this to: `We can free its bpobj_info_t, and if requested also free the now-empty bpo from disk and decrement its parent's subobj count.`

> Parent subobj(s) in our LIFO stack will need to have their directly          
> referenced BPs freed

I think that the parent bpo's would already have their directly referenced BP's freed, before any of their child bpo's are pushed on the stack.  The only bpo's on the stack that have directly referenced BP's would not have any children in the stack yet.  But maybe I've misunderstood what you mean by "parent" here?

To cover the `free` and non-`free` cases, maybe we should talk about `visiting` the BP's and bpo's rather than `freeing` them.  Also I think it's the callback that actually does the freeing of the referenced BP's (and there are some callbacks that don't free all BP's).

Reply:
Thanks.  I suppose if the code had been easy to write the comment would also be easy!

```
/*                                                                              
 * Given an initial bpo, start by freeing the BPs that are directly referenced  
 * by that bpo. If the bpo has subobjs, first re-push it to our LIFO stack so   
 * it can be revisited. Read in its last subobj and push the subobj to our      
 * stack. By popping items off our stack, eventually we will encounter          
 * a bpo that has no subobjs.  We can free its bpobj_info_t, and if             
 * requested also free the now-empty bpo from disk and decrement                
 * its parent's subobj count. We continue popping each subobj from our stack,   
 * re-pushing non-empty subobjs and visiting its last subobj until they too     
 * have no more subobjs, as so on.                                              
 */          ```